File: haskell/src/Solutions/Euler504.hs
module Solutions.Euler504 where
import Utils.Numeric(isSquare)

-- 1. Pick's theorem
-- 2. Given two lattice points A = (a, b) and X = (x, y),
--    there are gdc (x - a) (y - b) + 1 lattice points on it.
--    In code (+1) does not appear - we don't want to count ends twice.

f a b c d = let area2 = (a + c) * (b + d)
                boundPoints = gcd a b + gcd b c + gcd c d + gcd d a
            in isSquare $ (area2 + 2 - boundPoints) `div` 2

compute m = let l = [1..m] in length $ filter id [f a b c d | a <- l, b <- l, c <- l, d <- l]

euler504 :: IO String
euler504 = return $ show $ compute 100

File: rust/src/solutions/euler504.rs
use num::integer::{gcd, Roots};

fn f(a: u16, b: u16, c: u16, d: u16) -> bool {
    let area2 = (a + c) * (b + d);
    let bound_points = gcd(a, b) + gcd(b, c) + gcd(c, d) + gcd(d, a);
    let x = (area2 + 2 - bound_points) / 2;
    x.sqrt().pow(2) == x
}

fn compute(m: u16) -> usize {
    let mut ans = 0;
    for a in 1..=m {
        for b in 1..=m {
            for c in 1..=m {
                for d in 1..=m {
                    if f(a, b, c, d) {
                        ans += 1;
                    }
                }
            }
        }
    }
    ans
}

pub fn euler504() -> String {
    compute(100).to_string()
}
